systems/plants/TimeSteppingRigidBodyManipulator.m

```diff
@@ -300,34 +279,104 @@ function checkDirty(obj)
         % and implement equation (7) from Anitescu97, by collecting
         %   J = [JL; JP; n; D{1}; ...; D{mC}; zeros(nC,num_q)]
 
+        % a "possible" index is one that is close enough to contact or the
+        % limit to have a chance at being active.  now we don't even
+        % construct the M and w matrices for contacts/limits that seem
+        % impossible to reach within this timestep.  
+        possible_limit_indices = [];
+        possible_contact_indices = [];
+        possible_indices_changed = false;
+        
+        while (1) 
+        
+        if (nL > 0)
+          if (obj.position_control)
+            phiL = q(pos_control_index) - u;
+            JL = sparse(1:obj.manip.num_u,pos_control_index,1,obj.manip.num_u,obj.manip.num_positions);
+            phiL = [phiL;-phiL]; JL = [JL;-JL];
+            % dJ = 0 by default, which is correct here
+            dJL = zeros(length(phiL),num_q^2);
+          else
+            if (nargout>4)
+              [phiL,JL,dJL] = obj.manip.jointLimitConstraints(q);
+            else
+              [phiL,JL] = obj.manip.jointLimitConstraints(q);
+            end
+            if isempty(possible_limit_indices)
+              possible_limit_indices = (phiL + h*JL*qd) < z_inactive_guess_tol;
+            end
+            nL = sum(possible_limit_indices);
+            
+            % phi_check and J_check are the "impossible indices"
+            % which get checked at the end of the method (to make sure
+            % that they did not somehow make contact or hit the limit)
+            phi_check = phiL(~possible_limit_indices);
+            J_check = JL(~possible_limit_indices,:);
+            phiL = phiL(possible_limit_indices);
+            JL = JL(possible_limit_indices,:);
+            if (nargout>4)
+              dJL = dJL(possible_limit_indices,:);
+            end
+            if isempty(obj.LCP_cache.data.possible_limit_indices) || any(obj.LCP_cache.data.possible_limit_indices~=possible_limit_indices)
+              possible_indices_changed = true;
+            end
+            obj.LCP_cache.data.possible_limit_indices=possible_limit_indices;
+          end
+        else
+          phi_check = zeros(0,1);
+          J_check = zeros(0,num_q);
+          JL = zeros(0,num_q^2);
+        end
+        
         if (nContactPairs > 0)
           if (nargout>4)
             [phiC,normal,d,xA,xB,idxA,idxB,mu,n,D,dn,dD] = obj.manip.contactConstraints(q,true);
```

Comment:
i don't think we are relying on that.  multiple contact points support *should* work here once it's enabled.